x-pack/metricbeat/module/gcp/metrics/compute/metadata.go

```diff
@@ -180,44 +198,49 @@ func (s *metadataCollector) instanceZone(ts *monitoringpb.TimeSeries) string {
 	return ""
 }
 
-func (s *metadataCollector) getComputeInstances(ctx context.Context) {
-	if len(s.computeInstances) > 0 {
-		return
-	}
-
-	s.logger.Debug("Compute API Instances.AggregatedList")
+func (s *metadataCollector) fetchComputeInstances(ctx context.Context) (map[string]*computepb.Instance, error) {
+	s.logger.Info("Executing fetchComputeInstances via CacheRegistry request")
 
 	instancesClient, err := compute.NewInstancesRESTClient(ctx, s.opt...)
 	if err != nil {
-		s.logger.Errorf("error getting client from compute service: %v", err)
-		return
+		return nil, fmt.Errorf("error creating compute client: %w", err)
 	}
-
 	defer instancesClient.Close()
 
 	start := time.Now()
-	defer func() {
-		s.logger.Debugf("Total time taken for compute AggregatedList request: %s", time.Since(start))
-	}()
+	s.logger.Debug("Compute API Instances.AggregatedList starting...")
 
-	it := instancesClient.AggregatedList(ctx, &computepb.AggregatedListInstancesRequest{
+	req := &computepb.AggregatedListInstancesRequest{
 		Project: s.projectID,
-	})
+	}
+	it := instancesClient.AggregatedList(ctx, req)
+	fetchedInstances := make(map[string]*computepb.Instance)
 
+	pageCount := 0
+	instanceCount := 0
 	for {
 		instancesScopedListPair, err := it.Next()
 		if errors.Is(err, iterator.Done) {
 			break
 		}
-
 		if err != nil {
-			s.logger.Errorf("error getting next instance from InstancesScopedListPairIterator: %v", err)
-			break
+			s.logger.Errorf("Error fetching next instance page: %v", err)
+			return nil, fmt.Errorf("error iterating compute instances: %w", err)
 		}
+		pageCount++
 
 		instances := instancesScopedListPair.Value.GetInstances()
+		if instances == nil {
+			continue // Skip zones/regions with no instances
+		}
+
 		for _, instance := range instances {
-			s.computeInstances[instance.GetId()] = instance
+			instanceIdStr := strconv.FormatUint(instance.GetId(), 10)
+			fetchedInstances[instanceIdStr] = instance
+			instanceCount++
 		}
 	}
+
+	s.logger.Infof("Compute AggregatedList finished in %s. Fetched %d instances across %d pages.", time.Since(start), instanceCount, pageCount)
```

Comment:
nit : should this be `info` level ? Prior to the refactoring, it was a defer based `debug` level log isn't it ?

Reply:
good catch thanks!